File: ts/MergeSort.ts
const merge: (arr1:[], arr2:[]) => [] = (arr1, arr2) => {
    const results: [] = [];

    while(arr1.length && arr2.length){
        arr1[0] < arr2[0] ? results.push(arr1.shift()) : results.push(arr2.shift());
    }

    return results;
};


const mergeSort: (arr: []) => [] = arr => {
    if(arr.length === 1) {
        return arr;
    }

    const center: number = Math.floor(arr.length / 2);
    const left = arr.slice(0, center);
    const right = arr.slice(center);

    return merge(mergeSort(left), mergeSort(right));
}


console.log(mergeSort([5,1,3,2,4]));

File: python/merge_sort.py
import math


def merge(arr1, arr2):
    sorted = []

    while len(arr1) != 0 and len(arr2) != 0:
        if arr1[0] < arr2[0]:
            sorted.append(arr1.pop(0))
        else:
            sorted.append(arr2.pop(0))

    return [*sorted, *arr1, *arr2]


def merge_sort(arr):
    if len(arr) < 2:
        return arr

    halfPoint = math.floor(len(arr) / 2)
    left = arr[:halfPoint]
    right = arr[halfPoint:]

    return merge(merge_sort(left), merge_sort(right))


list_to_sort = [5, 3, 2, 1, 4]

list_to_merge = [7, 2, 4, 1, 10]


print(merge_sort(list_to_sort))
